compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -1225,6 +1232,61 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                     ty::ClauseKind::TypeOutlives(outlives).upcast(tcx),
                 ));
 
+                // Require that all AFIT will return something that can be coerced into `dyn*`
+                // -- a shim will be responsible for doing the actual coercion to `dyn*`.
+                if let Some(principal) = data.principal() {
+                    for supertrait in
+                        elaborate::supertraits(tcx, principal.with_self_ty(tcx, source))
+                    {
+                        if tcx.is_trait_alias(supertrait.def_id()) {
+                            continue;
+                        }
+
+                        for &assoc_item in tcx.associated_item_def_ids(supertrait.def_id()) {
+                            if !tcx.is_impl_trait_in_trait(assoc_item) {
+                                continue;
+                            }
+
+                            let mut bound_vars = supertrait.bound_vars().to_vec();
+                            let args = supertrait.skip_binder().args.extend_to(
+                                tcx,
+                                assoc_item,
+                                |arg, _| match arg.kind {
+                                    GenericParamDefKind::Lifetime => {
+                                        let kind = ty::BoundRegionKind::Named(
+                                            arg.def_id,
+                                            tcx.item_name(arg.def_id),
+                                        );
+                                        bound_vars.push(ty::BoundVariableKind::Region(kind));
+                                        ty::Region::new_bound(tcx, ty::INNERMOST, ty::BoundRegion {
+                                            var: ty::BoundVar::from_usize(bound_vars.len() - 1),
+                                            kind,
+                                        })
+                                        .into()
+                                    }
+                                    GenericParamDefKind::Type { .. }
+                                    | GenericParamDefKind::Const { .. } => unreachable!(),
+                                },
+                            );
```

Comment:
I think this, plus wrapping it in a binder again is better as a separate function, followed by mapping the binder below to turn it into a trait ref with a projection arg. Easier to follow the bound vars, and mapping generic params to additional bound vars feels like an operation that, even if never used again, is better as an isolated operation with a named function.

Reply:
I pulled this out into its own fn and left a doc comment explaining what it's doing